src/datascience-ui/interactive-common/mainStateController.ts

```diff
@@ -616,6 +616,16 @@ export class MainStateController implements IMessageHandler {
         }
     }
 
+    public renderUpdate(newState: {}) {
+        // This method should be called during the render stage of anything
+        // using this state Controller. That's because after shouldComponentUpdate
+        // render is next and at this point the state has been set.
+        // See https://reactjs.org/docs/react-component.html
+        // Otherwise we set the state in the callback during setState and this can be
+        // too late for any render code to use the stateController.
+        this.state = { ...this.state, ...newState };
```

Comment:
Am I reading this wrong or does this apply the state update twice on state adjusting operations?

Reply:
It will apply it twice but the second should be a dupe. It's in case the state is updated and a render doesn't occur.

---
In reply to: [320515042](https://github.com/IanMatthewHuff/vscode-python/pull/3#discussion_r320515042) [](ancestors = 320515042)

Reply:
I don't love that :(. But I don't have a better alternative here. I agree that it should be a dupe.

Reply:
It just updates the internal 'state' object. It doesn't cause any other side effects. It's really just a matter of timing as to when the internal state is changed after calling setState

---
In reply to: [320517087](https://github.com/IanMatthewHuff/vscode-python/pull/3#discussion_r320517087) [](ancestors = 320517087)